### ISSUE TITLE
Loss of Deployment Agent

### ISSUE THREAD

**Issue by cboulva99** [Microsoft/vsts-agent#1737]
## Have you tried trouble shooting?
No related information found in troubleshooting

## Agent Version and Platform
2.136.1 on Windows 2016 Version 1607

## VSTS Type and Version
VisualStudio.com

http://datagalaxy.visualstudio.com

## What's not working?
Loss of connection/files few hours after installation.

Experiences multiple times (reinstalled multiple times), after installation the deployment agent, using the provivded script in the Deployment Group tab (https://datagalaxy.visualstudio.com/DataGalaxy/_machinegroup?view=MachineGroupView&mgid=9&tab=Details), with or without integrated PAT in the installation, I get the same issue: after a few hours running perfectly well, with successful deployments to IIS, the agent stops working:

The configuration files are "automatically" deleted (.agent, .service, .credentials and .credentials_rsaparams files disappear), some of the binaries are also automatically removed: the sub folders in bin (powershell and localization folders), and also the externals subfolder.

I do have 401 errors all the time the agent runs properly, maybe it has something to do with the problem?

## Agent and Worker's Diagnostic Logs
Extract of the log:

**At install:**

[2018-07-30 09:57:30Z INFO AgentProcess] Agent is built for Windows - win-x64.
[2018-07-30 09:57:30Z INFO AgentProcess] RuntimeInformation: Microsoft Windows 10.0.14393 .
[2018-07-30 09:57:30Z INFO AgentProcess] Version: 2.136.1
[2018-07-30 09:57:30Z INFO AgentProcess] Commit: 961d0fd5f1cc92df469d266ad131a6bdd8df7d8a

**While running:**

[2018-07-30 09:57:34Z INFO HostContext] Well known config file 'RSACredentials': 'C:\vstsagent\A1\.credentials_rsaparams'
[2018-07-30 09:57:34Z INFO RSAEncryptedFileKeyManager] Loading RSA key parameters from file C:\vstsagent\A1\.credentials_rsaparams
[2018-07-30 09:57:34Z INFO Terminal] WRITE LINE: Connecting to the server.
[2018-07-30 09:57:34Z INFO MessageListener] Attempt to create session.
[2018-07-30 09:57:34Z INFO MessageListener] Connecting to the Agent Server...
[2018-07-30 09:57:34Z INFO VisualStudioServices] Starting operation Location.GetConnectionData
[2018-07-30 09:57:34Z WARN VisualStudioServices] Authentication failed with status code 401.
Cache-Control: no-cache
Date: Mon, 30 Jul 2018 09:57:34 GMT
Pragma: no-cache
P3P: CP="CAO DSP COR ADMa DEV CONo TELo CUR PSA PSD TAI IVDo OUR SAMi BUS DEM NAV STA UNI COM INT PHY ONL FIN PUR LOC CNT"
WWW-Authenticate: Basic realm="https://tfsprodweusu4.app.visualstudio.com/", Bearer, TFS-Federated
X-TFS-ProcessId: 20997a74-c94f-4709-be9e-3ae855d5c556
Strict-Transport-Security: max-age=31536000; includeSubDomains
ActivityId: 6f71ed26-30dd-4c1c-9204-5d237a9237f3
X-TFS-Session: 90f901c9-0b91-4a6f-a2fc-7eea86028ad8
X-VSS-E2EID: 64748482-20f5-416e-90b2-99affa9faee9
X-FRAME-OPTIONS: SAMEORIGIN
X-TFS-FedAuthRealm: https://tfsprodweusu4.app.visualstudio.com/
X-TFS-FedAuthIssuer: https://datagalaxy.visualstudio.com/
X-VSS-AuthorizationEndpoint: https://datagalaxy.vssps.visualstudio.com/
X-VSS-ResourceTenant: 00000000-0000-0000-0000-000000000000
X-AspNet-Version: 4.0.30319
X-VSS-S2STargetService: 00000002-0000-8888-8000-000000000000/visualstudio.com
X-TFS-FedAuthRedirect: https://app.vssps.visualstudio.com/_signin?realm=datagalaxy.visualstudio.com&reply_to=https%3A%2F%2Fdatagalaxy.visualstudio.com%2F_apis%2FconnectionData%3FconnectOptions%3D0%26lastChangeId%3D174657579%26lastChangeId64%3D174657579&redirect=1&context=eyJodCI6MiwiaGlkIjoiOWI5MzI2ZDItMTE5MC00ZTdkLWI2OTItOThjODZmMTY3MGI1IiwicXMiOnt9LCJyciI6IiIsInZoIjoiIiwiY3YiOiIiLCJjcyI6IiJ90#ctx=eyJTaWduSW5Db29raWVEb21haW5zIjpbImh0dHBzOi8vbG9naW4ubWljcm9zb2Z0b25saW5lLmNvbSIsImh0dHBzOi8vbG9naW4ubWljcm9zb2Z0b25saW5lLmNvbSJdfQ2
X-Powered-By: ASP.NET
X-Content-Type-Options: nosniff
X-MSEdge-Ref: Ref A: 2F72A67CE75C47B0BD43174E053EA949 Ref B: AMS04EDGE0721 Ref C: 2018-07-30T09:57:34Z

[2018-07-30 09:57:34Z INFO RSAEncryptedFileKeyManager] Loading RSA key parameters from file C:\vstsagent\A1\.credentials_rsaparams
[2018-07-30 09:57:34Z INFO RSAEncryptedFileKeyManager] Loading RSA key parameters from file C:\vstsagent\A1\.credentials_rsaparams

**Later On:**

[2018-07-30 11:34:15Z INFO MessageListener] No message retrieved from session '58e42d46-33c9-4903-97f7-09e24d9b45af' within last 30 minutes.
[2018-07-30 11:48:27Z WARN VisualStudioServices] Authentication failed with status code 401.
Cache-Control: no-cache
Date: Mon, 30 Jul 2018 11:48:27 GMT
Pragma: no-cache
P3P: CP="CAO DSP COR ADMa DEV CONo TELo CUR PSA PSD TAI IVDo OUR SAMi BUS DEM NAV STA UNI COM INT PHY ONL FIN PUR LOC CNT"
WWW-Authenticate: Basic realm="https://tfsprodweusu4.app.visualstudio.com/", Bearer, TFS-Federated
X-TFS-ProcessId: eae058d6-5c69-450b-be0a-24cacba56018
Strict-Transport-Security: max-age=31536000; includeSubDomains
ActivityId: 7a881bb1-e829-43a2-aa5f-8f90a373372d
X-TFS-Session: 90f901c9-0b91-4a6f-a2fc-7eea86028ad8
X-VSS-E2EID: 2bf63f3b-3a5e-4081-93e5-44a5fb44dc14
X-FRAME-OPTIONS: SAMEORIGIN
X-TFS-FedAuthRealm: https://tfsprodweusu4.app.visualstudio.com/
X-TFS-FedAuthIssuer: https://datagalaxy.visualstudio.com/
X-VSS-AuthorizationEndpoint: https://datagalaxy.vssps.visualstudio.com/
X-VSS-ResourceTenant: 00000000-0000-0000-0000-000000000000
X-AspNet-Version: 4.0.30319
X-VSS-S2STargetService: 00000002-0000-8888-8000-000000000000/visualstudio.com
X-TFS-FedAuthRedirect: https://app.vssps.visualstudio.com/_signin?realm=datagalaxy.visualstudio.com&reply_to=https%3A%2F%2Fdatagalaxy.visualstudio.com%2F_apis%2Fdistributedtask%2Fpools%2F6%2Fmessages%3FsessionId%3D58e42d46-33c9-4903-97f7-09e24d9b45af%26lastMessageId%3D29&redirect=1&context=eyJodCI6MiwiaGlkIjoiOWI5MzI2ZDItMTE5MC00ZTdkLWI2OTItOThjODZmMTY3MGI1IiwicXMiOnt9LCJyciI6IiIsInZoIjoiIiwiY3YiOiIiLCJjcyI6IiJ90#ctx=eyJTaWduSW5Db29raWVEb21haW5zIjpbImh0dHBzOi8vbG9naW4ubWljcm9zb2Z0b25saW5lLmNvbSIsImh0dHBzOi8vbG9naW4ubWljcm9zb2Z0b25saW5lLmNvbSJdfQ2
X-Powered-By: ASP.NET
X-Content-Type-Options: nosniff
X-MSEdge-Ref: Ref A: 2FA1DC46CC8E4CF79742915A22B98DE5 Ref B: AMS04EDGE0709 Ref C: 2018-07-30T11:48:27Z

[2018-07-30 11:48:27Z INFO RSAEncryptedFileKeyManager] Loading RSA key parameters from file C:\vstsagent\A1\.credentials_rsaparams

**And then, it breaks up:**

[2018-07-31 01:37:38Z WARN VisualStudioServices] Authentication failed with status code 401.
Cache-Control: no-cache
Date: Tue, 31 Jul 2018 01:37:38 GMT
Pragma: no-cache
P3P: CP="CAO DSP COR ADMa DEV CONo TELo CUR PSA PSD TAI IVDo OUR SAMi BUS DEM NAV STA UNI COM INT PHY ONL FIN PUR LOC CNT"
WWW-Authenticate: Basic realm="https://tfsprodweusu4.app.visualstudio.com/", Bearer, TFS-Federated
X-TFS-ProcessId: 3b7c3e6f-33cd-4746-9673-502e3c4543d5
Strict-Transport-Security: max-age=31536000; includeSubDomains
ActivityId: 1036648d-e915-4728-a9e1-b3f686c4107b
X-TFS-Session: 90f901c9-0b91-4a6f-a2fc-7eea86028ad8
X-VSS-E2EID: 8da6e8e2-db7d-4b66-b8bc-3b6e214e0309
X-FRAME-OPTIONS: SAMEORIGIN
X-TFS-FedAuthRealm: https://tfsprodweusu4.app.visualstudio.com/
X-TFS-FedAuthIssuer: https://datagalaxy.visualstudio.com/
X-VSS-AuthorizationEndpoint: https://datagalaxy.vssps.visualstudio.com/
X-VSS-ResourceTenant: 00000000-0000-0000-0000-000000000000
X-AspNet-Version: 4.0.30319
X-VSS-S2STargetService: 00000002-0000-8888-8000-000000000000/visualstudio.com
X-TFS-FedAuthRedirect: https://app.vssps.visualstudio.com/_signin?realm=datagalaxy.visualstudio.com&reply_to=https%3A%2F%2Fdatagalaxy.visualstudio.com%2F_apis%2Fdistributedtask%2Fpools%2F6%2Fmessages%3FsessionId%3D58e42d46-33c9-4903-97f7-09e24d9b45af%26lastMessageId%3D39&redirect=1&context=eyJodCI6MiwiaGlkIjoiOWI5MzI2ZDItMTE5MC00ZTdkLWI2OTItOThjODZmMTY3MGI1IiwicXMiOnt9LCJyciI6IiIsInZoIjoiIiwiY3YiOiIiLCJjcyI6IiJ90#ctx=eyJTaWduSW5Db29raWVEb21haW5zIjpbImh0dHBzOi8vbG9naW4ubWljcm9zb2Z0b25saW5lLmNvbSIsImh0dHBzOi8vbG9naW4ubWljcm9zb2Z0b25saW5lLmNvbSJdfQ2
X-Powered-By: ASP.NET
X-Content-Type-Options: nosniff
X-MSEdge-Ref: Ref A: 7F66EF61D06A4E49BF45DA2313C5AF99 Ref B: AMS04EDGE0709 Ref C: 2018-07-31T01:37:38Z

[2018-07-31 01:37:38Z ERR  MessageListener] Catch exception during get next message.
[2018-07-31 01:37:38Z ERR  MessageListener] System.Security.Cryptography.CryptographicException: RSA key file C:\vstsagent\A1\.credentials_rsaparams was not found
   at Microsoft.VisualStudio.Services.Agent.Listener.Configuration.RSAEncryptedFileKeyManager.GetKey()
   at Microsoft.VisualStudio.Services.WebApi.VssSigningCredentials.RSASigningToken.GetSignature(Byte[] input)
   at Microsoft.VisualStudio.Services.WebApi.Jwt.JsonWebToken.GetSignature(JWTHeader header, JWTPayload payload, JWTAlgorithm alg, VssSigningCredentials signingCredentials)
   at Microsoft.VisualStudio.Services.WebApi.Jwt.JsonWebToken.Create(String issuer, String audience, DateTime validFrom, DateTime validTo, DateTime issuedAt, IEnumerable`1 additionalClaims, JsonWebToken actor, String actorToken, VssSigningCredentials credentials, Boolean allowExpiredCertificate)
   at Microsoft.VisualStudio.Services.WebApi.Jwt.JsonWebToken.Create(String issuer, String audience, DateTime validFrom, DateTime validTo, IEnumerable`1 additionalClaims, VssSigningCredentials credentials)
   at Microsoft.VisualStudio.Services.OAuth.VssOAuthJwtBearerAssertion.GetBearerToken()
   at Microsoft.VisualStudio.Services.OAuth.VssOAuthJwtBearerClientCredential.SetParameters(IDictionary`2 parameters)
   at Microsoft.VisualStudio.Services.OAuth.VssOAuthTokenHttpClient.CreateRequestContent(IVssOAuthTokenParameterProvider[] parameterProviders)
   at Microsoft.VisualStudio.Services.OAuth.VssOAuthTokenHttpClient.<GetTokenAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.OAuth.VssOAuthTokenProvider.<OnGetTokenAsync>d__14.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.Common.IssuedTokenProvider.GetTokenOperation.<GetTokenAsync>d__21.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.Common.IssuedTokenProvider.<GetTokenAsync>d__33.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.Common.VssHttpMessageHandler.<SendAsync>d__17.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.Common.VssHttpRetryMessageHandler.<SendAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at System.Net.Http.HttpClient.<FinishSendAsyncBuffered>d__58.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.<SendAsync>d__51.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.<SendAsync>d__47`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.<SendAsync>d__28`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.Agent.Listener.MessageListener.<GetNextMessageAsync>d__13.MoveNext()
[2018-07-31 01:37:38Z INFO MessageListener] Retriable exception: RSA key file C:\vstsagent\A1\.credentials_rsaparams was not found

## Related Repositories
Please ensure you are logging issues to the correct repository in order to get the best support.

- [Tasks Repository](https://github.com/Microsoft/vsts-tasks) - contains all of the inbox tasks we ship with VSTS/TFS. If you are having issues with tasks in Build/Release jobs (e.g. unreasonable task failure) please log an issue [here](https://github.com/Microsoft/vsts-tasks/issues).
- [Hosted Agent Image Repository](https://github.com/Microsoft/vsts-image-generation) - contains the VM image used in the VSTS Hosted Agent Pool. If you are having Build/Release failures that seems like they are related to software installed on the Hosted Agent (e.g. the DotnetSDK is missing or the AzureSDK is not on the latest version) please log an issue [here](https://github.com/Microsoft/vsts-image-generation/issues).

If you are hitting a generic issue about VSTS/TFS, please report it to the [Developer Community](https://developercommunity.visualstudio.com/spaces/21/index.html)

**Comment by TingluoHuang** [comment]
@cboulva99 how was your VM/machine or agent got deployed? are you using any Azure ARM template/extension?

the 401 will happen about every 50 mins, the agent use a session token to communicate with service, the token valid for 50 mins, so every 50 mins, the agent needs to renew the token.

**Comment by cboulva99** [comment]
The VM was installed manually. I tried to install the Azure ARM for VSTS Deployment agent but the installation failed. Maybe I need to clean this from the VM? I finally installed manually the Agent on my two servers using the script provided in the VSTS Deployment Group Details tab.

**Comment by cboulva99** [comment]
Finally, the problem is solved!
It was the attempt that I made of installing the Deployment agent using the AzureRM Extension  which was left at a state of Provisioning failed. It seems this conflicted with the manually installed agent, and removed the configuration files every night around midnight. As soon as I removed the failed provisioning, the disappearing configuration files issue was gone!

Thank you @TingluoHuang for the explanation of the 401, and for your mention of the Azure ARM extension.
